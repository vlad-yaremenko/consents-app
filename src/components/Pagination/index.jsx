import React, { useState } from 'react';
import { Pagination } from '@material-ui/lab';

export const withPagination = (WrappedComponent, perPage) => {
  const Component = (props) => {
    const [page, setPage] = useState(1);
    const [data, setData] = useState(props.data || []);
    const [pages, setPages] = useState(Math.floor(props.data.length / perPage));

    React.useEffect(() => {
      setData(props.data.slice((page - 1) * perPage).slice(0, perPage));
      setPages(Math.ceil(props.data.length / perPage));
    }, [props.data, page]);

    return (
      <>
        <WrappedComponent {...props} data={data} />

        <Pagination count={pages} page={page} onChange={(e, page) => setPage(page)} />
      </>
    );
  };
  Component.displayName = `WithPagination${WrappedComponent.displayName}`;

  return Component;
}
