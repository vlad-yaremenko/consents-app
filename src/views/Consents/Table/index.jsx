import React from 'react';
import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow
} from '@material-ui/core';

const Component = (props) => {
  const decorateAgreeTo = (arr) => {
    return arr.join(', ')
  };

  return (
    <TableContainer component={Paper}>
      <Table aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell>Name</TableCell>
            <TableCell>Email</TableCell>
            <TableCell align="right">Agree to</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {props.data.map(row => (
            <TableRow key={row.id}>
              <TableCell>{row.name}</TableCell>
              <TableCell>{row.email}</TableCell>
              <TableCell align="right">{decorateAgreeTo(row.agreeTo)}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  )
};
Component.displayName = 'Table';

export default Component;
