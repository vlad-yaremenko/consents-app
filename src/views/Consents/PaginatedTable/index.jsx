import { withPagination } from '@components/Pagination';
import Table from '../Table';

const Component = (props) => {
  const perPage = 2;
  return withPagination(Table, perPage)(props);
}

export default Component;
