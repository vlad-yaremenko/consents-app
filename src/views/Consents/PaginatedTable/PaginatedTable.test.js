import { render } from '@testing-library/react';
import PaginatedTable from './index';
import { withPagination } from '@components/Pagination';
import Table from '../Table';

jest.mock('../Table', () => jest.fn());
jest.mock('@components/Pagination', () => ({
  withPagination: jest.fn()
}));

beforeEach(() => {
  Table.mockImplementation(() => <table data-testid="table"></table>);
  withPagination.mockImplementation((component) => (props) => <div data-testid="pagination">{component(props)}</div>);
});

afterEach(() => {
  jest.resetAllMocks();
});

test('should render table wrapped by pagination', () => {
  const { container, getByTestId } = render(<PaginatedTable />);

  expect(getByTestId('table')).toBeInTheDocument();
  expect(getByTestId('pagination')).toBeInTheDocument();

  expect(container).toMatchSnapshot();
});

test('should set per page', () => {
  render(<PaginatedTable />);

  const PER_PAGE = 2;
  expect(withPagination).toHaveBeenCalledWith(expect.anything(), PER_PAGE);
});
