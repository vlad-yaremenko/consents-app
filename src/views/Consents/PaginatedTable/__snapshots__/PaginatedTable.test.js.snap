// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render table wrapped by pagination 1`] = `
<div>
  <div
    data-testid="pagination"
  >
    <table
      data-testid="table"
    />
  </div>
</div>
`;
